ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs

```diff
@@ -0,0 +1,188 @@
+
+import           Ouroboros.Network.Block (ChainHash, StandardHash)
+import qualified Ouroboros.Network.Block as Block
+
+import           Ouroboros.Storage.ChainDB.API
+
+import           Ouroboros.Consensus.Ledger.Abstract
+import           Ouroboros.Consensus.Mempool.API
+import           Ouroboros.Consensus.Util (repeatedly)
+
+{-------------------------------------------------------------------------------
+  Top-level API
+-------------------------------------------------------------------------------}
+
+openMempool :: (MonadSTM m, StandardHash blk, ApplyTx blk)
+            => ChainDB m blk hdr -> m (Mempool m blk)
+openMempool chainDB = do
+    env <- initMempoolEnv chainDB
+    return Mempool {
+        addTxs = implAddTxs env
+      , getTxs = implGetTxs env
+      }
+
+{-------------------------------------------------------------------------------
+  Internal state
+-------------------------------------------------------------------------------}
+
+-- | Internal state in the mempool
+data InternalState blk = IS {
+     -- | Transactions currently in the mempool
+      isTxs :: Seq (GenTx blk)
+
+      -- | The tip of the chain that 'isTxs' was validated against
+    , isTip :: ChainHash blk
+    }
+
+data MempoolEnv m blk hdr = MempoolEnv {
+      mpEnvChainDB  :: ChainDB m blk hdr
+    , mpEnvStateVar :: TVar m (InternalState blk)
+    }
+
+initInternalState :: InternalState blk
+initInternalState = IS Seq.empty Block.GenesisHash
+
+initMempoolEnv :: MonadSTM m => ChainDB m blk hdr -> m (MempoolEnv m blk hdr)
+initMempoolEnv chainDB = do
+    isVar <- atomically $ newTVar initInternalState
+    return $ MempoolEnv chainDB isVar
+
+{-------------------------------------------------------------------------------
+  Implementation
+-------------------------------------------------------------------------------}
+
+-- TODO: This may return some transactions as invalid that aren't new. Remove?
+implAddTxs :: forall m blk hdr. (MonadSTM m, StandardHash blk, ApplyTx blk)
+           => MempoolEnv m blk hdr
+           -> [GenTx blk]
+           -> m [(GenTx blk, LedgerError blk)]
+implAddTxs mpEnv@MempoolEnv{..} txs = atomically $ do
+    ValidationResult{..} <- validateNew <$> validateIS mpEnv
+    writeTVar mpEnvStateVar IS { isTxs = vrValid
+                               , isTip = vrBefore
+                               }
+    return vrInvalid
+  where
+    validateNew :: ValidationResult blk ->  ValidationResult blk
+    validateNew = extendsVR $ map (NotPrevApplied, ) txs
+
+implGetTxs :: (MonadSTM m, StandardHash blk, ApplyTx blk)
+           => MempoolEnv m blk hdr
+           -> STM m (Seq (GenTx blk ))
+implGetTxs mpEnv@MempoolEnv{..} = do
+    ValidationResult{..} <- validateIS mpEnv
+    -- TODO (maybe): log the transactions that we silently discard
+    writeTVar mpEnvStateVar IS { isTxs = vrValid
+                               , isTip = vrBefore
+                               }
+    return vrValid
```

Comment:
So answering my Q above: the impl strategy here is to validate on demand, and in fact to do so every time. So this is a relatively expensive operation.

Reply:
Re-validating is expensive, but mind that we have a cheap check to see if we _need_ to, and once we do, we cache the results. This seemed an easy solution that would nonetheless give quite strong guarantees, without doing more work than we need to.